x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredModelSnapshotsRemover.java

```diff
@@ -157,116 +163,84 @@ protected void removeDataBefore(
             listener.onResponse(true);
             return;
         }
-        LOGGER.debug("Considering model snapshots of job [{}] that have a timestamp before [{}] for removal", job.getId(), cutoffEpochMs);
-
-        SearchRequest searchRequest = new SearchRequest();
-        searchRequest.indices(AnomalyDetectorsIndex.jobResultsAliasedName(job.getId()));
-
-        QueryBuilder activeSnapshotFilter = QueryBuilders.termQuery(
-            ModelSnapshotField.SNAPSHOT_ID.getPreferredName(), job.getModelSnapshotId());
-        QueryBuilder retainFilter = QueryBuilders.termQuery(ModelSnapshot.RETAIN.getPreferredName(), true);
-        QueryBuilder query = createQuery(job.getId(), cutoffEpochMs)
-            .filter(QueryBuilders.existsQuery(ModelSnapshot.SNAPSHOT_DOC_COUNT.getPreferredName()))
-            .mustNot(activeSnapshotFilter)
-            .mustNot(retainFilter);
-
-        SearchSourceBuilder source = new SearchSourceBuilder();
-        source.query(query);
-        source.size(MODEL_SNAPSHOT_SEARCH_SIZE);
-        source.sort(ModelSnapshot.TIMESTAMP.getPreferredName());
-        source.fetchSource(false);
-        source.docValueField(Job.ID.getPreferredName(), null);
-        source.docValueField(ModelSnapshotField.SNAPSHOT_ID.getPreferredName(), null);
-        source.docValueField(ModelSnapshot.TIMESTAMP.getPreferredName(), "epoch_millis");
-        searchRequest.source(source);
-        searchRequest.setParentTask(getParentTaskId());
+        LOGGER.debug(() -> new ParameterizedMessage(
+            "Considering model snapshots of job [{}] that have a timestamp before [{}] for removal",
+            job.getId(),
+            cutoffEpochMs));
 
         long deleteAllBeforeMs = (job.getModelSnapshotRetentionDays() == null)
             ? 0 : latestTimeMs - TimeValue.timeValueDays(job.getModelSnapshotRetentionDays()).getMillis();
-        client.execute(SearchAction.INSTANCE, searchRequest, new ThreadedActionListener<>(LOGGER, threadPool,
-            MachineLearning.UTILITY_THREAD_POOL_NAME, expiredSnapshotsListener(job.getId(), deleteAllBeforeMs, listener), false));
+        ActionListener<QueryPage<ModelSnapshot>> snapshotsListener = expiredSnapshotsListener(job, deleteAllBeforeMs, listener);
+        jobResultsProvider.modelSnapshots(job.getId(),
```

Comment:
I was running into async callback hell. But, if we just use the underlying `modelSnapshots` request from here and then filter out the assigned snapshot later, we reduce the number of calls and it flows more cleanly.